### PR TITLE
Use wash interface to remake a batch heat items menu

### LINKED ISSUE #72477: Adding ways to heat food in batch
### Is your feature request related to a problem? Please describe.

Cooking tools that use electricity, such as electric stoves, can only heat one portion of food at a time with the heat food function.
When it comes to 20 portions of food that need to be heated or defrosted, the process of performing 20 heating actions can become very painful.

### Solution you would like.

If you can use an electric stove to heat a pot that holds food, a container of this type that has a cooking function (or is made of metal), put the item need to be heated into the pot, and heat pot up, could be better.
To make this happen：
1. Adding holsters to the electric stove，and put pot on it to heat.
2. When use stove, make some of the pot selective.

### Describe alternatives you have considered.

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by yuganxia**
#### Summary
Interface "Use wash interface to remake a batch heat menu"

#### Purpose of change
resolve #72477 

#### Describe the solution

Use wash_items function and interface as basic frame, make a new way of heating items, include liquid and solid items.
Make heat procedure can be done by batch.The heating procedure should under watch so you can't idle it.
Frozen item takes more time on heating, _Frozen liquid in container should melt_(This part is not done yet.)

Active objtect:
Base on COOK and BOIL qualities to seperate tools into 3 class:
1. heat_liquid_items: BOIL 1-2 COOK 0
2. heat_solid_items: COOK 1
3. heat_all_items: COOK 2-3 
4. Hotplate flag 1,2

Container:
When active tools, 
1. If it have action of heat_solid_items and don't have container, such like spear and tongs, you can only use it to heat one solid item a time, but that item have no volume limit.
2. if it have action of heat_liquid_items and heat_all_items, you must to empty it to heat, or will heat items in it ,you can heat items no more than max container volume a time.
3. if item has action of hotplate you can use it as heat source, or not, you need select a heat source.

Heat source:
1.Hotplate can only use itself as heat source, hotplate 2 can used as heat source for other tools.
2.If near fire, use fire as heat source.
3.If near tools has hotplate 2 or funiture with heating function on it, and still have charge, use it as heat source. 
Calculate ammo used by mass.

Consume:
Calculate time and fuel needed by weight of selected items.

#### Testing
It just work.
image
image
image
image

### CODE REVIEW

**Comment by yuganxia** [comment]
1. I just don't know how to make tools linked up to elect grid or car to be functional.
2. action part is not done yet.

UI is functional now. Added extra needed for FROZEN item,will take 3 times time and fuel to heat up.
I just notice that HOTPLATE tool don't have CONTAINER pocket.I don't know why.

**Comment by yuganxia** [comment]
Add qualities HOTPLATE 1 and 2 as heat source, 
HOTPLATE 1 means it can be used as heat source for itself, such as coffee maker or concrete mixer.
HOTPLATE 2 means it can be used for pots.

**Comment by yuganxia** [comment]
Basic function should be work right now..Not sure about UPS.JSON is not changed yet.
After finished function check maybe i will add another pull to do that part.
Original HEAT_FOOD function has 2 time multi on FROZEN item so change mine from 3 to 2.

**Comment by yuganxia** [comment]
Some of function not done yet, such as melt frozen liquid in container and put heated item in container.Might do it later(or not, this functions might comes with bugs or make function exterm complex).

File: src/iuse.cpp
Comment (Qrox) [inline src/iuse.cpp:8330]:
```suggestion
                {{ _( "Fuel" ), ammo }},
```

File: src/iuse.cpp
Comment (anothersimulacrum) [inline src/iuse.cpp:8121]: These functions have a return value that you ignore. If you intend to use their return value use, otherwise don't have them return a value.
See this comment for information on what that means.
https://github.com/CleverRaven/Cataclysm-DDA/blob/cd18044375172e446488292dfe1f3a6e31934ce8/src/iuse.h#L24-L28

File: src/iuse.h
Comment (anothersimulacrum) [inline src/iuse.h:87]: These are helper functions - they should be `static` functions with `iuse.cpp`, and only return `std::optional<int>` if they need to.

File: src/iuse.cpp
Comment (anothersimulacrum) [inline src/iuse.cpp:8246]: There's no reason for this to be done in the helper functions instead of the iuse.

File: src/iuse.cpp
Comment (anothersimulacrum) [inline src/iuse.cpp:8248]: This will be an incorrectly constructed item location for items with the `ALLOWS_REMOTE_USE` flag, if those items are used remotely.

File: src/iuse.cpp
Comment (anothersimulacrum) [inline src/iuse.cpp:8271]: You constructed heater the exact same way above. Why are you doing it again and again down here?

File: src/iuse.cpp
Comment (anothersimulacrum) [inline src/iuse.cpp:8276]: Not all hotplate analogues run on batteries.

File: src/iuse.cpp
Comment (anothersimulacrum) [inline src/iuse.cpp:8315]: This whole section could (and should) be moved to a separate helper function, `find_heater` (or similar).
You can return a struct with the information you're populating here.

File: src/iuse.cpp
Comment (anothersimulacrum) [inline src/iuse.cpp:8181]: Delete this function, and delete heat_items...

File: src/iuse.cpp
Comment (anothersimulacrum) [inline src/iuse.cpp:8242]: ..and call this function, with a new argument that only allows one item to be selected.

File: src/iuse.cpp
Comment (anothersimulacrum) [inline src/iuse.cpp:8401]: This needs to happen at the end of the activity, not at the start.

This prevents the activity being cancelled and still consuming energy.
It's also worth adding a do_turn to the activity to ensure that there still is enough energy/the conditions are right for the heating to continue. If they're not, the activity should be cancelled, with no effect.

Cancellation could result in completing as much heating as was possible for the time elapsed/energy available, but this is not necessary and is complicated for an already big PR.

File: src/iuse.cpp
Comment (anothersimulacrum) [inline src/iuse.cpp:8091]: Not sure this PR needs to tackle this, but the energy requirements for this are definitely wrong. @anoobindisguise may have some advice there.

**Comment by yuganxia** [comment]
I still don't know what to do with atomic heaters, adding HOTPLATE,3 and 4 might fix it, but it doesn't quite make sense with 3>2 but 3 can't use as parent heat sorce. And heating pads, maybe we just keep this.

File: src/activity_actor.cpp
Comment (anothersimulacrum) [inline src/activity_actor.cpp:7604]: If there's no heater, no energy.
```suggestion
    int available_heater = 0;
```

**Review by anothersimulacrum — COMMENTED** [review]
More review coming later, typo fix and small behavior correction right now.

**Comment by GuardianDll** [comment]
Is it ready to merge?

**Comment by yuganxia** [comment]
> Is it ready to merge?

I'm waiting for a review from [anothersimulacrum](https://github.com/anothersimulacrum).

**Comment by Maleclypse** [comment]
> > Is it ready to merge?
> 
> I'm waiting for a review from @anothersimulacrum Main function is done. Some of the details need to be fix, cost formula and atomic stuff.

It might be time to assume anothersim is unable to get back to this currently.  What is needed to complete this besides their review?

**Comment by yuganxia** [comment]
> > > Is it ready to merge?
> > 
> > 
> > I'm waiting for a review from @anothersimulacrum Main function is done. Some of the details need to be fix, cost formula and atomic stuff.
> 
> It might be time to assume anothersim is unable to get back to this currently. What is needed to complete this besides their review?

Nothing I think, I corrected the formula part, atomic part may be I'll open a new PR later.
Yeah, it's good to go.

File: src/activity_actor.cpp
Comment (anothersimulacrum) [inline src/activity_actor.cpp:7658]: In testing, the heated items got duplicated. I suspect that's because you're adding them here, but not deleting them before doing this. It's probably acceptable to skip this and just heat them in-place?

**Review by anothersimulacrum — COMMENTED** [review]
Sorry for the long delay, and thank you for putting up with it :)

It would be nice to combine all the other heating iuses to follow this one path, but I agree that's too much for this PR.

File: src/activity_actor.cpp
Comment (anothersimulacrum) [reply src/activity_actor.cpp:7668]: Did you fail to push it? It doesn't seem corrected.

Comment (yuganxia) [reply src/activity_actor.cpp]: > Did you fail to push it? It doesn't seem corrected.

Indeed, Fixed now.

**Review by anothersimulacrum — APPROVED** [review]
LGTM. There's a bunch of work to do transferring all the other heating methods to this one, but this one seems solid.

**Pull request merged**
